MLP recurrent

timmeinhardt opened the pull request:
This PR consists of two commits related to recurrent policy models.

The first commit implements the recurrent option for the MLP policy. This results in multiple methods shared by the MLP and CNN model which I extracted in a `NNBase` class.

The second commit doesn't add any new functionality but suggests a new naming convention for _observations_, _current obervations_ and _states_. From my understanding, a general naming convention in RL is that _states_ is an umbrella term for _observations_. They can be the same but if for example `num_stack=4` one state is made up of 4 observations. With the term _states_ the current code refers to the hidden state of the recurrent layer of a policy. As this might be confusing, I suggest a new naming convention with `current_obs`  to `states` and `states` to `recurrent_hidden_states`. Inside the policy class the recurrent hidden states are abbreviated to `rnn_hxs`.

ikostrikov commented:
Hey! Looks good! I will review it over the weekend.

ikostrikov commented:
Renaming states to recurrent_hidden_states sounds good. Since we don't know true states but approximate them with RNNs.

But for current_obs I would prefer to keep it as observations. Since we might not know true states. And the naming convention is already used in gym.

timmeinhardt commented:
I was not aware of the naming convention in gym. And considering the unknown true states I agree that we should keep the _obs_ and _current_obs_. But the differentiation between states and recurrent network hidden states should definitely be made.

ikostrikov commented:
Ok. Looks good to me.

Thanks a lot!

Could you please rename the variables? Let's discuss naming conventions in a separate PL/issue.

Also let me know if you have time to resolve the conflicts.

timmeinhardt commented:
I removed the conflicts and _obs_ to _states_ renaming. Also I changed observations to obs to be consistent everywhere and pluralised the arguments of `Storage.insert`. I think the confusion between the plural of the `Storage.obs` variable and the `obs` argument is smaller than always pass pluralised variables and then the arguments are singular.

ikostrikov commented:
Ok. Looks good, I'm ready to merge.

Let me know if you have time to resolve the conflicts.

timmeinhardt commented:
I added another commit which introduces an optional `base_kwargs` argument to the `Policy` class. This allows to easily change parameters of the base model (e.g., `hidden_size`) from the `main.py` file.

timmeinhardt commented:
And another commit in which I introduce a `is_recurrent` method to the policy and it’s base. This cleans up the code base (no `hasattr` calls) and could make it easier to try different types of recurrent  cells, e.g. LSTMs.
